### PR TITLE
feat: process ChainLocks embedded in connected blocks

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
## Issue being fixed or feature implemented

A synced node that misses a ChainLock message currently stores the signature embedded in later coinbases without using it to update its best ChainLock. This change lets it recover the ChainLock when those blocks connect.

## What was done?

- Process coinbase signatures from the block already supplied to the block-connected notification, after v20 activation and while ChainLocks are enabled and the node is synced.
- Derive the signed ancestor as `containing height - bestCLHeightDiff - 1`, checking the unsigned offset before conversion. Reuse normal quorum-signature verification, duplicate suppression, asynchronous enforcement, and inventory relay.
- Adapt the original implementation to the current `chainlock::ChainlockHandler`. The peer manager supplies the quorum manager and handles the returned relay inventory from its block-connected callback.
- Replace arithmetic-only tests with coverage that calls the production handler. Replace the previous functional test with a two-node, single-member-quorum test that disconnects the regular node and delivers blocks exclusively through `submitblock`. Cover zero/nonzero offsets, restart recovery, repeated signatures, and normal relay after reconnection.
- Add `doc/release-notes-6236.md`.

The old nine-commit series was consolidated and rebased onto `552577d32a` (develop). Existing upstream ChainLock tests and quorum helpers are retained. The unrelated `CCoinbaseChainlock` pair-to-class refactor, unused helpers, and duplicate tests are omitted; extraction still uses the in-memory block, without a disk read. No mining, serialization, or coinbase consensus-validation rules are changed.

## How Has This Been Tested?

Locally on Apple Silicon macOS, using the prebuilt depends prefix and a wallet-enabled build without GUI:

- Full `make -j8` build.
- `test_dash --run_test=llmq_chainlock_tests`.
- `feature_llmq_chainlocks_automatic.py`, `feature_llmq_chainlocks.py`, and `rpc_verifychainlock.py`.
- Mutation check: restoring the original missing `-1` makes the new unit test fail at all three tested offsets and makes the functional test fail when the isolated receiver must recover its first coinbase-delivered ChainLock. Both pass with the fix restored.
- Python lint with the repository-pinned mypy 0.981, including its 312-file type-check pass; circular dependencies, includes, assertions, whitespace, and formatting of changed C++ lines.

### CI follow-up

The first full run passed lint, the standard Linux/SQLite/no-wallet/ASan test jobs, and the other platform builds. Two failures in unchanged code are being retried:

- [Multiprocess build](https://github.com/dashpay/dash/actions/runs/34502431059/job/102958988687): `net_processing.cpp` compilation started at 16:58:31 before the recursive BLS build generated `relic_conf.h` at 16:58:36. The missing generated-header dependency is present on the base revision.
- [TSan test](https://github.com/dashpay/dash/actions/runs/34502431059/job/102966811603): `feature_protx_version.py` detected concurrent first-use initialization of `ep2_curve_get_s3` in unchanged `src/dashbls/src/legacy.cpp`, reached by two DKG signing threads. This is a real pre-existing race, not a timeout or a ChainLock-test failure; a retry does not fix the underlying BLS issue.

Only failed jobs were rerun in attempt 2. No test suppression or unrelated source change was added.

## Breaking Changes

None. Nodes can now learn an existing, valid ChainLock from a connected block as well as from a standalone message.

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: test/functional/feature_llmq_chainlocks_automatic.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+'''
+
+import time
+from test_framework.test_framework import DashTestFramework
+from test_framework.util import assert_equal, assert_raises_rpc_error, force_finish_mnsync, wait_until_helper
+
+
+class LLMQChainLocksAutomaticTest(DashTestFramework):
+    def set_test_params(self):
+        self.set_dash_test_params(6, 5)
```
Comment (knst) [inline test/functional/feature_llmq_chainlocks_automatic.py]: consider ab42708dac315e367ae75e9b9654bfa7b986b6be

firstly, you don't need to test masternode functionality, only just "some" chainlocks. Single-node-quorum speed up this functional test significantly.

Secondly, you should not isolate masternode for this functional test, because it adds instability in case if this masternode is in quorum. Instead, isolate regular node.

It takes only 18 seconds to run with my changes.

Comment (knst) [reply test/functional/feature_llmq_chainlocks_automatic.py]: or maybe even  71db9ad7bb51398c02cfe3d3ba0f27a2c797051f

it uses 1 less node and just 10 seconds to run

**Comment by thepastaclaw** [comment]
✅ Final review complete — no blockers (commit 47b0bfb5) · triage: normal

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Code Review

The PR refactors std::pair<CBLSSignature, uint32_t> into CCoinbaseChainlock cleanly, but the new automatic-detection path in CChainLocksHandler::BlockConnected has an off-by-one in the height derivation that diverges from every other use site (specialtxman, miner, rawtransaction, functional test). The result is that the constructed CChainLockSig points to the wrong block and the signature cannot verify, making the new feature non-functional. The new tests do not catch the bug — the unit tests assert tautological arithmetic and the functional test never forces the coinbase-only delivery path.

🔴 1 blocking | 🟡 2 suggestion(s) | 💬 3 nitpick(s)

*Reviewed commit: 50eb7dbd21*

<details>
<summary>1. [BLOCKING] src/llmq/chainlocks.cpp:395 — Off-by-one in clsig_height derivation breaks automatic chainlock detection</summary>

`bestCLHeightDiff` is encoded such that the signed block lives at `containing_block_height - bestCLHeightDiff - 1`. This convention is used at every other call site:

- `src/evo/specialtxman.cpp:71`: `int curBlockCoinbaseCLHeight = pindex->nHeight - static_cast<int>(cbTx.bestCLHeightDiff) - 1;`
- `src/node/miner.cpp:151`: `int prevCLHeight = pindexPrev->nHeight - static_cast<int>(prevBlockCoinbaseChainlock->heightDiff) - 1;`
- `src/rpc/rawtransaction.cpp:643`: `pTipBlockIndex->GetAncestor(pTipBlockIndex->nHeight - cbtx_best_cl->heightDiff - 1)`
- Existing functional helper in `test_coinbase_best_cl`: `best_cl_height = cb_height - best_cl_height_diff - 1`

The new code drops the `-1`, so the constructed `CChainLockSig` carries a height/hash pair that the BLS signature does not actually cover. `ProcessNewChainLock` will fail signature verification for every legitimate coinbase-embedded chainlock, rendering the new automatic-detection feature non-functional. Note that the new unit tests at lines 180, 233, 295, 325 mirror the same incorrect formula (e.g. `block_height - height_diff` without `-1`), and the new functional test does not exercise the BlockConnected path, so CI does not catch the bug.

```suggestion
        int32_t clsig_height = pindex->nHeight - static_cast<int32_t>(coinbase_cl.heightDiff) - 1;
```

</details>

<details>
<summary>2. [SUGGESTION] src/llmq/chainlocks.cpp:398 — Tighten clsig_height bounds check to strictly less than pindex->nHeight</summary>

With the correct formula `clsig_height = pindex->nHeight - heightDiff - 1`, the signed block is strictly an ancestor of `pindex`, so the upper bound should be `>= pindex->nHeight`, not `> pindex->nHeight`. The current `clsig_height > pindex->nHeight` permits `clsig_height == pindex->nHeight`, which (after the off-by-one fix) cannot legitimately occur and indicates this path was never exercised. The fact that the existing check is `>` rather than `>=` is itself a tell that the off-by-one was masked during testing.

```suggestion
        if (clsig_height < 0 || clsig_height >= pindex->nHeight) {
```

</details>

<details>
<summary>3. [SUGGESTION] test/functional/feature_llmq_chainlocks_automatic.py:50 — Functional test does not exercise the BlockConnected coinbase-CL path</summary>

The test mines blocks and waits for chainlocks via `wait_for_chainlocked_block` / `wait_for_chainlocked_block_all_nodes`, then inspects coinbase fields. On this single-MN setup chainlocks always arrive through the normal LLMQ signing / gossip flow in `CChainLocksHandler` — independent of the new `BlockConnected` code path. The edge-case sub-test isolates node 1 (a non-MN) but later calls `self.sync_blocks()`, which re-enables the gossip path, so the new node still receives the chainlock via normal channels. As a result the test passes even though the off-by-one bug in chainlocks.cpp:395 makes coinbase-derived chainlocks unverifiable.

Additionally, line 92 computes `cl_height = height - cbtx["bestCLHeightDiff"]` (without `-1`), which is the same incorrect formula as the production bug — so even the test's own assertion (`verifychainlock(cl_block_hash, ...)`) would fail to confirm correctness; it's wrapped in `try/except` that swallows failures, so it logs success regardless.

To actually validate the feature: start (or restart) a node that cannot receive chainlocks via gossip (e.g. disconnected from the MN before any CLSIG message arrives, then fed only blocks whose coinbases carry CLs) and assert via `getbestchainlock` that the new node ends up with the chainlock. Compute the expected height as `height - heightDiff - 1` to match the on-the-wire convention.

</details>

<details>
<summary>4. [NITPICK] src/test/llmq_chainlock_tests.cpp:180 — Several new unit tests assert tautologies, not behavior</summary>

`automatic_chainlock_detection_logic_test` (180), `automatic_chainlock_edge_cases_test` (233), `chainlock_ancestor_lookup_edge_cases_test` (295), and `chainlock_comparison_and_validation_test` (325) compute integer arithmetic locally (`1000 - 5 == 995`, `600 > 500`, etc.) and assert the result of the local computation. They never invoke `BlockConnected`, `GetCoinbaseChainlock`, or any other production function. They also reinforce the incorrect `block_height - height_diff` formula (no `-1`), so they would not have surfaced the off-by-one in chainlocks.cpp:395. Replace with tests that construct a `CBlock` with a coinbase `CCbTx`, hand it to `GetCoinbaseChainlock`, and assert the derived height matches the on-the-wire convention used by the miner and validation code.

</details>

<details>
<summary>5. [NITPICK] src/evo/cbtx.cpp:260 — Redundant V20 deployment check in GetNonNullCoinbaseChainlock</summary>

`GetNonNullCoinbaseChainlock` checks `DeploymentActiveAt(..., DEPLOYMENT_V20)` at line 267, then calls `GetCoinbaseChainlock` which repeats the identical check at line 238. Either drop the outer check or have `GetNonNullCoinbaseChainlock` operate on the already-fetched block via the inner helper. Not a correctness issue.

</details>

<details>
<summary>6. [NITPICK] src/test/util/llmq_tests.h:27 — Unused test helpers added in this PR</summary>

`GetLLMQParams`, `CreateValidCommitment`, `CreateChainLock`, `CreateBitVector`, and `TestSerializationRoundtrip` are added but never called by any new test in this PR (the new tests only consume `CreateRandomBLSSignature`, `GetTestQuorumHash`, `GetTestBlockHash`). One reference exists at `src/test/llmq_chainlock_tests.cpp:76`, but only as a comment noting `TestSerializationRoundtrip` is broken and skipped. Trim the unused helpers, or land them in the PR where they are actually used.

</details>

_Inline review posting hit GitHub HTTP 422, so I posted the same verified findings as a top-level review body._

**Review by thepastaclaw — APPROVED** [review]
## Final validation — Phase 1 + Phase 2

The implementation correctly extracts valid ChainLock signatures from connected coinbase transactions only when synchronization, ChainLocks, and V20 activation prerequisites are satisfied. It validates the height offset before conversion, derives the signed ancestor safely, and reuses the existing ChainLock verification, duplicate suppression, enforcement, and relay path; the accompanying unit and functional tests cover the relevant recovery and relay scenarios.

Source: reviewer 1: `gemini-3.8-flash-high` (agent: `phase1-reviewer`, role: `general`); reviewer 2: `gemini-3.8-flash-high` (agent: `phase1-reviewer`, role: `dash-core-commit-history`); reviewer 3: `gpt-6-astra` (agent: `phase2-reviewer`, role: `general`); reviewer 4: `gpt-6-astra` (agent: `phase2-reviewer`, role: `dash-core-commit-history`); final verifier: `gpt-6-astra` (agent: `astra-verifier`, role: `final-verifier`)

### Review provenance
- Triage: `normal` by `gpt-6-astra` (effort low) — The change adds and wires ChainLock recovery logic across the handler, peer manager, unit tests, and functional tests, but it does not alter consensus validation, cryptographic primitives, peer-facing deserialization, funds movement, or persistence migrations.
- Phase 1 reviewers: `gemini-3.8-flash-high` — general (completed, effort high); agent `phase1-reviewer`, `gemini-3.8-flash-high` — dash-core-commit-history (completed, effort high); agent `phase1-reviewer`
- Phase 1 model: `gemini-3.8-flash-high` — antigravity quota: weekly 84% left, 5h 93% left
- Fresh verifier: `gpt-6-astra` — final-verifier; agent `astra-verifier`
- Phase 2 reviewers: `gpt-6-astra` — general (completed, effort high); agent `phase2-reviewer`, `gpt-6-astra` — dash-core-commit-history (completed, effort high); agent `phase2-reviewer`

File: src/chainlock/handler.cpp [hunk 2]
```diff
@@ -98,6 +101,35 @@ void ChainlockHandler::UpdateTxFirstSeenMap(const Uint256HashSet& tx, const int6
     }
 }
 
+MessageProcessingResult ChainlockHandler::ProcessCoinbaseChainLock(const CBlock& block, const CBlockIndex* pindex,
+                                                                   const llmq::CQuorumManager& qman)
+{
+    if (!m_mn_sync.IsBlockchainSynced() || !m_chainlocks.IsEnabled() || pindex == nullptr ||
+        !DeploymentActiveAt(*pindex, Params().GetConsensus(), Consensus::DEPLOYMENT_V20) || block.vtx.empty()) {
```
Comment (knst) [inline src/chainlock/handler.cpp]: nit: why not check block's ccbtx's version instead of DeploymentActiveAt?
